src/coreclr/vm/method.cpp

```diff
@@ -2485,526 +2636,219 @@ MethodDesc* MethodDesc::GetMethodDescFromStubAddr(PCODE addr, BOOL fSpeculative
-
-    CentralJumpCode* pCode = (CentralJumpCode*)p;
-    memcpy(pCode, &c_CentralJumpCode, TEP_CENTRAL_JUMP_SIZE);
-    pCode->Setup (GetPreStubCompactARMEntryPoint(), this);
-
-    _ASSERTE(p + TEP_CENTRAL_JUMP_SIZE == (BYTE*)temporaryEntryPointsWriterHolder.GetRW() + size);
-
-#else // TARGET_ARM
-
-    _ASSERTE(p == (BYTE*)temporaryEntryPointsWriterHolder.GetRW() + size);
-
-#endif // TARGET_ARM
-
-    ClrFlushInstructionCache((LPVOID)temporaryEntryPoints, size);
-
-    SetHasCompactEntryPoints();
-    return temporaryEntryPoints;
 }
-#endif // !DACCESS_COMPILE
-
-#endif // HAS_COMPACT_ENTRYPOINTS
 
 //*******************************************************************************
-PCODE MethodDescChunk::GetTemporaryEntryPoint(int index)
+void MethodDescChunk::DetermineAndSetIsEligibleForTieredCompilation()
 {
-    LIMITED_METHOD_CONTRACT;
+    WRAPPER_NO_CONTRACT;
 
-#ifdef HAS_COMPACT_ENTRYPOINTS
-    if (HasCompactEntryPoints())
+    if (!DeterminedIfMethodsAreEligibleForTieredCompilation())
     {
-#ifdef TARGET_ARM
-
-        return GetTemporaryEntryPoints() + COMPACT_ENTRY_ARM_CODE + THUMB_CODE + index * TEP_ENTRY_SIZE;
-
-#else // TARGET_ARM
-
-        int fullBlocks = index / TEP_MAX_BLOCK_INDEX;
-        int remainder = index % TEP_MAX_BLOCK_INDEX;
+        int count = GetCount();
 
-        return GetTemporaryEntryPoints() + 1 + (fullBlocks * TEP_FULL_BLOCK_SIZE) +
-            (remainder * TEP_ENTRY_SIZE) + ((remainder >= TEP_MAX_BEFORE_INDEX) ? TEP_CENTRAL_JUMP_SIZE : 0);
-
-#endif // TARGET_ARM
-    }
-#endif // HAS_COMPACT_ENTRYPOINTS
+        // Determine eligibility for tiered compilation
+        {
+            MethodDesc *pMD = GetFirstMethodDesc();
+            bool chunkContainsEligibleMethods = pMD->DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk();
 
-    return Precode::GetPrecodeForTemporaryEntryPoint(GetTemporaryEntryPoints(), index)->GetEntryPoint();
-}
+    #ifdef _DEBUG
+            // Validate every MethodDesc has the same result for DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk
+            MethodDesc *pMDDebug = GetFirstMethodDesc();
+            for (int i = 0; i < count; ++i)
+            {
+                _ASSERTE(chunkContainsEligibleMethods == pMDDebug->DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk());
+                pMDDebug = (MethodDesc *)(dac_cast<TADDR>(pMDDebug) + pMDDebug->SizeOf());
+            }
+    #endif
+            if (chunkContainsEligibleMethods)
+            {
+                for (int i = 0; i < count; ++i)
+                {
+                    if (pMD->DetermineAndSetIsEligibleForTieredCompilation())
+                    {
+                        _ASSERTE(pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk());
+                    }
+                    else
+                    {
+                        _ASSERTE(!pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk());
+                    }
```

Comment:
```suggestion
                    _ASSERTE(pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk() == pMD->DetermineAndSetIsEligibleForTieredCompilation());
```

Reply:
That is not a safe change. The `DetermineAndSetIsEligibleForTieredCompilation` method must run in release builds.

Reply:
Oh, I didn't see the "AndSet". Sorry. I would then update this slightly.

```suggestion
                    bool isEligible = pMD->DetermineAndSetIsEligibleForTieredCompilation();
                    _ASSERTE(isEligible == pMD->IsEligibleForTieredCompilation_NoCheckMethodDescChunk());
```